### PR TITLE
sink(cdc): fallback when preparing statements meets error

### PR DESCRIPTION
**Pull request by hicqu**
### What problem does this PR solve?

Issue Number: ref #8508 

### What is changed and how it works?

Fallback to normal SQL executor when preparing statements meets error.

#### Tests 

 - No code

### Release note 

```release-note
None
```

### CODE REVIEW

File: cdc/sink/dmlsink/txn/mysql/mysql.go [hunk 1]
```diff
@@ -656,23 +653,25 @@ func (s *mysqlBackend) sequenceExecute(
...
-				}
 
+		var prepStmt *sql.Stmt
+		if s.cachePrepStmts {
+			if stmt, ok := s.stmtCache.Get(query); ok {
+				prepStmt = stmt.(*sql.Stmt)
+			} else if stmt, err := s.db.Prepare(query); err == nil {
+				prepStmt = stmt
 				s.stmtCache.Add(query, stmt)
+			} else {
```
Comment (asddongmen) [inline cdc/sink/dmlsink/txn/mysql/mysql.go:664]: Could you please add a comment to tell why we need to do this in the second else block?

**Review by asddongmen — APPROVED** [review]

File: cdc/sink/dmlsink/txn/mysql/mysql.go [hunk 2]
```diff
@@ -138,13 +137,11 @@ func NewMySQLBackends(
 		// Because each connection can not hold at lease one prepared statement.
 		if maxPreparedStmtCount == 0 || maxPreparedStmtCount/(cfg.WorkerCount+1) == 0 {
 			cachePrepStmts = false
-		} else if maxPreparedStmtCount/(cfg.WorkerCount+1) < prepStmtCacheSize {
-			// if maxPreparedStmtCount/(cfg.WorkerCount+1) < prepStmtCacheSize,
-			// it means that the prepared statement cache is too large on clientsize.
-			// adjust the size of the prepared statement cache on clientsize.
-			// to avoid error `Can't create more than max_prepared_stmt_count statements`
-			prepStmtCacheSize = maxPreparedStmtCount / (cfg.WorkerCount + 1)
```
Comment (amyangfei) [inline cdc/sink/dmlsink/txn/mysql/mysql.go:146]: https://github.com/pingcap/tiflow/pull/8387 has tried to fixed related issue, seems the `prepStmtCacheSize` is not set to a proper value.

Comment (hicqu) [reply cdc/sink/dmlsink/txn/mysql/mysql.go:146]: The solution can't resolve the situation with more than one CDC instances. So I remove `cacheSize`. Now cache evict is triggered by errors returned by `db.Prepare`.

File: cdc/sink/dmlsink/txn/mysql/mysql.go [hunk 3]
```diff
@@ -122,29 +127,28 @@ func NewMySQLBackends(
...
 		}
+	}
+
+	var stmtCache *lru.Cache
+	if cachePrepStmts {
 		stmtCache, err = lru.NewWithEvict(prepStmtCacheSize, func(key, value interface{}) {
```
Comment (dbsid) [inline cdc/sink/dmlsink/txn/mysql/mysql.go:152]: I think we can remove the const prepStmtCacheSize, and replace it with maxPreparedStmtCount

File: cdc/sink/dmlsink/txn/mysql/mysql.go [hunk 4]
```diff
@@ -122,29 +127,28 @@ func NewMySQLBackends(
...
-	var stmtCache *lru.Cache
 	if cachePrepStmts {
 		// query the size of the prepared statement cache on serverside
 		maxPreparedStmtCount, err := pmysql.QueryMaxPreparedStmtCount(ctx, db)
 		if err != nil {
 			return nil, err
 		}
+		if maxPreparedStmtCount == -1 {
+			// NOTE: seems TiDB doesn't follow MySQL's specification.
+			maxPreparedStmtCount = math.MaxInt
```
Comment (dbsid) [inline cdc/sink/dmlsink/txn/mysql/mysql.go:138]: Should we change `math.MaxInt` to 1048576?

Comment (hicqu) [reply cdc/sink/dmlsink/txn/mysql/mysql.go:138]: I guess it's unnecessary. Even if in MySQL we still can set it bigger than 1048576.
So `-1` just means nolimit, so let's just keep it.

File: cdc/sink/dmlsink/txn/mysql/mysql.go (hunk 2 above)
Comment (dbsid) [reply cdc/sink/dmlsink/txn/mysql/mysql.go:146]: although this does not fix the situation with more than one CDC instances. It set the reasonable limit for the cache size. This value is much lower than `16*1024` when there is more than one worker count. If we hard code the cache size with `16*1024`, we might need to keep try db.Prepare() then got error and fallback to query.

**Review by dbsid — APPROVED** [review]

**Review by amyangfei — APPROVED** [review]

**Pull request merged**